### PR TITLE
Implementation of Username Recovery feature.

### PR DESCRIPTION
**Pull request by GDRDABARERA**

### CODE REVIEW

File: product/carbon-home/conf/identity/profile-mapping.yaml [hunk 1]
```diff
@@ -123,4 +123,30 @@ profiles:
...
+   required: true
+   readonly: false
+   regex: "*"
+   dataType: "text"
+  -
+   claimURI: "http://wso2.org/claims/email"
+   displayName: "Email"
+   required: true
+   readonly: false
+   regex: "*"
```
Comment (omindu) [inline product/carbon-home/conf/identity/profile-mapping.yaml]: Use a suitable regex for email address.

File: product/carbon-home/conf/identity/profile-mapping.yaml [hunk 2]
```diff
@@ -123,4 +123,31 @@ profiles:
...
+   dataType: "text"
+  -
+   claimURI: "http://wso2.org/claims/email"
+   displayName: "Email"
+   required: true
+   readonly: false
+   regex: "^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*
```
Comment (omindu) [inline product/carbon-home/conf/identity/profile-mapping.yaml]: The given regex pattern doesn't work. You can use the following.
`^([a-zA-Z0-9_\.\-])+\@(([a-zA-Z0-9\-])+\.)+([a-zA-Z0-9]{2,4})+$`

File: product/carbon-home/conf/identity/domain-config.yaml [hunk 3]
```diff
@@ -139,4 +139,4 @@ domains:
         databaseType: "MySQL"
         hashAlgorithm: "SHA256"
         keyLength: 256
-        iterationCount: 4096
+        iterationCount: 4096
```
Comment (ashensw) [inline product/carbon-home/conf/identity/domain-config.yaml]: Keep the newline.

File: portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/pages/recovery/username.hbs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                                {{/if}}
+                            </form>
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </div>
+    </div>
+
+{{/fillZone}}
```
Comment (ashensw) [inline portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/pages/recovery/username.hbs]: Add a new line.

Comment (GDRDABARERA) [reply portal/apps/org.wso2.carbon.iam.user-portal.app/src/main/pages/recovery/username.hbs]: Fixed with https://github.com/wso2/product-is/pull/1371

File: product/carbon-home/conf/identity/profile-mapping.yaml (hunk 1 above)
Comment (GDRDABARERA) [reply product/carbon-home/conf/identity/profile-mapping.yaml]: This is used in the all config, this is only working. Need to refactor all in the next implementation

File: product/carbon-home/conf/identity/domain-config.yaml (hunk 3 above)
Comment (GDRDABARERA) [reply product/carbon-home/conf/identity/domain-config.yaml]: Fixed with https://github.com/wso2/product-is/pull/1371

**Pull request merged**